File: leetcode/two pointer/1004. Max Consecutive Ones III/js/main.js
/* 
  Найди самый длинный подмассив с максимум K нулями.
  
  Алгоритм: Sliding Window
  
  У нас есть два указателя start = 0 и end = 0

  Двигаем указатель end вправо пока у нас соблюдается условие ограничения.
  
  И при этом после каждой итерации подсчитываем максимум разности end - start.

  Если условие ограничения не выполняется попадаем во вложенный цикл.

  В котором двигаем указатель start до тех пор пока выходим за рамки ограничения.

  После цикла опять подсчитываем максимум end - start

          L ->        R
  nums = [0 1 1 1 0 1 0 1]    k = 2

*/

// Time O(N)
// Space O(1)
const longestOnes = (nums, k) => {
  let start = 0;
  let end = 0;
  let cnt = 0;
  let maxLen = 0;

  while (end < nums.length) {
    if (nums[end] == 0) {
      cnt++;
    }

    end++;

    while (cnt > k) {
      if (nums[start] == 0) {
        cnt--;
      }
      start++;
    }

    if (maxLen < end - start) {
      maxLen = end - start;
    }
  }

  return maxLen;
};

File: leetcode/two pointer/1004. Max Consecutive Ones III/py/main.py
class Solution:
    # Time O(N)
    # Space O(1)
    def longestOnes(self, A, K):
        n = len(A)
        start, end = 0, 0
        max_length = 0
        cnt = 0

        while end < n:
            if A[end] == 0:
                cnt += 1
            end += 1

            while cnt > K:
                if A[start] == 0:
                    cnt -= 1
                start += 1

            if max_length < end - start:
                max_length = end - start

        return max_length

    # Time O(N^2)
    # Space O(1)
    def longestOnes_II(self, A, K):
        n = len(A)
        ans = 0

        for i in range(n):
            cnt = K

            if A[i] == 0:
                cnt -= 1

            j = i + 1

            while j < n and cnt >= 0:
                if A[j] == 0:
                    cnt -= 1
                j += 1

            if cnt < 0:
                j -= 1

            ans = max(ans, j - i)

        return ans